File: Lab 9/t3.c
#include <stdio.h>
#include <string.h>

#define BUFSIZE 4096

#define _STR(s) #s
#define STR(s) _STR(s)

void flush_stdin (void)
{
  int c = '\0';

  // NOTE: getchar will block stdin.
  while ((c = getchar()) && c != '\n' && c != EOF) {}
}

int main (void)
{
  char buf1[BUFSIZE] = {0};
  char buf2[BUFSIZE] = {0};
  int r = 0;

  for (;;) {
    // Always flush stdin before reading it.
    // flush_stdin ();

    // Round 1
    puts ("Please set your password:");
    r = scanf ("%" STR(BUFSIZE) "s", buf1);
    if (r != 1) {
      puts ("Illegal input.");
      return 1;
    }

    // Length >= 8
    if (strlen (buf1) < 8) {
      puts ("Password not long enough.");
      flush_stdin ();
      continue;
    }

    // Per requirement: use the first 20 characters.
    if (strlen (buf1) > 20) {
      buf1[20] = '\0';
    }

    // Always flush stdin before reading it.
    flush_stdin ();

    // Round 2
    puts ("Please confirm your password:");
    r = scanf ("%" STR(BUFSIZE) "s", buf2);
    if (r != 1) {
      puts ("Illegal input.");
      return 1;
    }

    // Per requirement: use the first 20 characters.
    if (strlen (buf2) > 20) {
      buf2[20] = '\0';
    }

    // Check if two inputs match
    if (memcmp (buf1, buf2, BUFSIZE) != 0) {
      puts ("Passwords don't match.");
      flush_stdin ();
      continue;
    } else {
      printf ("Your password \"%s\" is confirmed.\n", buf2);
      break;
    }
  }

  return 0;
}

File: Lab 9/t3.go
package main

import (
  "fmt"
  "os"
  "bufio"
  "strings"
)

func main() {
  var (
    pw1, pw2 string
    buf []byte // For ReadLine use
    e error
  )

  stdin := bufio.NewReader(os.Stdin)

  for {
    // Always flush stdin before reading it.
    stdin.Discard(stdin.Buffered())

    // Round 1
    fmt.Println("Please set your password:")
    buf, _, e = stdin.ReadLine()
    if e != nil {
      fmt.Println("Read error:", e)
      return
    }

    // Length >= 8
    // NOTE: The builtin `len` will count bytes instead of runes.
    // NOTE: strings.Count returns 1 + the number of Unicode code points in s.
    // if len(buf1) < 8 {
    if strings.Count(string(buf), "") < 9 {
      fmt.Println("Password not long enough.")
      continue
    }
    pw1 = string(buf)

    // Always flush stdin before reading it.
    stdin.Discard(stdin.Buffered())

    // Round 2
    fmt.Println("Please confirm your password:")
    buf, _, e = stdin.ReadLine()
    if e != nil {
      fmt.Println("Read error:", e)
      return
    }
    pw2 = string(buf)

    // Check if two inputs match
    if pw1 != pw2 {
      fmt.Println("Passwords don't match.")
      continue
    } else {
      fmt.Println("Your password", pw1, "is confirmed.")
      break
    }
  }
}
